app/javascript/views/ResetPassword.tsx

```diff
@@ -0,0 +1,163 @@
+import React from 'react';
+import { inject, observer } from 'mobx-react';
+import SessionStore, { RequestState } from '../stores/session_store';
+import { Header, Form, Message } from 'semantic-ui-react';
+import { RouterStore } from 'mobx-react-router';
+import DbSqlIcon from '../shared/DbSqlIcon';
+import { RouteComponentProps } from 'react-router';
+import { resetPassword as resetPasswordCall } from '../api/user';
+
+interface MatchParams {
+  id: string;
+}
+interface ResetPasswordProps extends RouteComponentProps<MatchParams> { }
+
+interface InjectedProps extends ResetPasswordProps {
+  sessionStore: SessionStore;
+  routerStore: RouterStore;
+}
+
+enum PasswordState { None, InvalidLength, NotEqual, Ok, InvalidRequest }
+
+@inject('sessionStore', 'routerStore')
+@observer
+export default class ResetPassword extends React.Component<ResetPasswordProps> {
+  state = {
+    passwordState: PasswordState.None,
+    resetState: null,
+    requestState: RequestState.None
+  };
+
+  private password: string = '';
+  private passwordConfirmation: string = '';
+  get injected() {
+    return this.props as InjectedProps;
+  }
+
+  get id() {
+    return this.props.match.params.id;
+  }
+
+  get resetToken() {
+    return this.queryParams.get('reset_token');
+  }
+
+  get queryParams() {
+    return new URLSearchParams(this.props.location.search);
+  }
+
+  resetPassword() {
+    if (!this.validatePassword()) return;
+
+    this.setState({ requestState: RequestState.Waiting });
+    resetPasswordCall(
+      this.id,
+      this.resetToken || '',
+      this.password,
+      this.passwordConfirmation
+    ).then(() => {
+      this.setState({ requestState: RequestState.Success });
+      this.injected.routerStore.push({
+        pathname: '/login',
+        search: '?reset=success'
+      });
+    }).catch((error) => {
+      this.setState({ requestState: RequestState.Error });
+      const msg = error.response.data.error || 'Unexpected server error';
+      this.setState({ resetState: msg });
+    });
+  }
+
+  validatePassword(): boolean {
+    if (this.password !== this.passwordConfirmation) {
+      this.setState({ passwordState: PasswordState.NotEqual });
+      return false;
+    }
+    if (this.password.length < 8 || this.password.length > 72) {
```

Comment:
it looks like this validation logic is duplicated (similar logic can be found in the form to update the password, create a user)